Trwalke/ui automation for mobile

trwalke opened the pull request:
Intrgrating Xamarin UI Test framework.
migrating the Azure Keyvault Access classes to the dev branch

Task (https://identitydivision.visualstudio.com/DevEx/_workitems/edit/480655)

Commit:
Added selenium framework.
Implemented acquireToken flow for android.

Commit:
Revert "Added selenium framework."

This reverts commit 082ea54c9c284c6d9d9639d128b1426672493b88.

Commit:
Included Xamarin UI test framework
Added wrapper for Xamarin UI test framework

Commit:
Adding key vault/Lab access
adding wrapper to xamarin UI testing framework

Commit:
Removed static from KeyVaultConfiguration.
changed android xforms app to use the UI Automation app registration.

Commit:
Merge remote-tracking branch 'origin/dev' into trwalke/UiAutomationForMobile

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs:
```diff
@@ -0,0 +1,34 @@
+
+            controller.Tap("secondPage");
+            controller.Tap("acquireToken");
+            controller.EnterText("i0116", user.Upn, true);
```
>EnterText [](start = 23, length = 9)

is there a faster way to insert text in a field? e.g. via copy/paste? when watching the automation running this seems to take forever. #Resolved

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs:
```diff
@@ -0,0 +1,118 @@
+            Tap(elementID, isWebElement, new TimeSpan(0, 0, waitTime));
+        }
+
+        public void EnterText(string elementID, string text, bool isWebElement)
```
>EnterText [](start = 20, length = 9)

I mentioned this in another place. For some of these text entering elements the automation takes forever. does something better exist in the test infra, to have the text inputs go faster (e.g. via copy/paste). #Resolved

trwalke replied:
@henrik-me  The good thing about the Xamarin UI test framework is that it does a copy and paste for text input by default when using "EnterText()". #Resolved

henrik-me replied:
Can the same be used for the desktop tests? #Resolved

trwalke replied:
With the current setup of the desktop tests, slow text input is the only option. However, with newer versions of the selenium framework, it may now be possible. ill investigate. #Resolved

henrik-me reviewed (changes requested):
:clock1:

henrik-me reviewed (approved):
:shipit:

Commit:
Changing keyvault authentication to use client credentials.

Downgrading nuget packages to resolve build errors

jennyf19 commented on adal/devApps/XFormsApp.Droid/XFormsApp.Droid.csproj:
```diff
@@ -31,10 +31,10 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <AndroidUseSharedRuntime>True</AndroidUseSharedRuntime>
+    <AndroidUseSharedRuntime>false</AndroidUseSharedRuntime>
```
I would do upper case here and line 37...goes w/the others #Resolved

jennyf19 commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -173,6 +186,7 @@ void LogCallback(LogLevel level, string message, bool containsPii)
             {
                 Device.BeginInvokeOnMainThread(() =>
                 {
+                    this.testResult.Text = string.IsNullOrWhiteSpace(accessToken) ? "Succsess: False" : "Succsess: True";
```
Check spelling on `Success` #Resolved

jennyf19 commented on adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        /// <summary>
+        /// Try and locate a certificate matching the given <paramref name="thumbprint"/> by searching in 
+        /// the in the <see cref="StoreName.My"/> store name for all all available <see cref="StoreLocation"/>s.
```
repeated "in the in the" and "all all available" #Resolved

jennyf19 commented on adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultConfiguration.cs:
```diff
@@ -0,0 +1,42 @@
+        /// The ID of the test harness client.
+        /// </summary>
+        /// <remarks>
+        /// This should be configured as as having access to the Key Vault instance specified at <see cref="Url"/>.
```
"as as" #Resolved

jennyf19 reviewed (commented):
this will be great @trwalke - excited to try it out

trwalke committed:
Update XFormsApp.Droid.csproj

Updating casing

trwalke committed:
Update SecondPage.cs

Changing spelling

trwalke committed:
Update CertificateHelper.cs

Grammar changes

trwalke committed:
Update KeyVaultConfiguration.cs

Grammar changes

trwalke committed:
Update CoreTests.cs

Changing spelling

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj:
```diff
@@ -1,13 +1,14 @@
     <RootNamespace>Test.ADAL.NET.UIAutomation</RootNamespace>
     <AssemblyName>Test.ADAL.NET.UIAutomation</AssemblyName>
-    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.1</TargetFrameworkVersion>
```
Why downgrade TargetFramework ? #Pending

trwalke replied:
Same as other comment
 #Resolved

henrik-me replied:
I can't find the other comment...

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj:
```diff
@@ -32,10 +33,10 @@
-    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="2.1.1" />
+    <PackageReference Include="Microsoft.Azure.KeyVault" Version="2.0.6" />
```
Why  downgrade these?

trwalke replied:
@henrik-me newer versions ran into compatibility issues that prevented the build from passing on VSTS

henrik-me replied:
compatibility or missing components on the box?

---
In reply to: [199915776](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1109#discussion_r199915776) [](ancestors = 199915776)

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -54,12 +56,12 @@ public class KeyVaultSecretsProvider
         {
             _config = new KeyVaultConfiguration();
-            _config.AuthType = KeyVaultAuthenticationType.UserCredential;
+            _config.AuthType = KeyVaultAuthenticationType.ClientCertificate;
```
has info on how to place the cert in the right location been added to the "how to setup a new build/test machine"?

trwalke replied:
Yes

bgavrilMS commented on adal/devApps/XFormsApp.Droid/Properties/AndroidManifest.xml:
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto" package="com.Microsoft.XFormsDroid">
```
Are these changes needed?

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs:
```diff
@@ -0,0 +1,20 @@
+			{
+				return ConfigureApp.Android.InstalledApp("com.Microsoft.XFormsDroid").StartApp();
+
+            }
```
Spacing..

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs:
```diff
@@ -0,0 +1,20 @@
+
+            }
+
+			return ConfigureApp.iOS.StartApp();
```
It'll be safer to add a switch statement over that Platform enum and to return an app for each supported platform and a NotImplemented or NotSupported exception for the other ones.

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Test.ADAL.NET.UIAutomation
+{
+	public class AppInitializer
```
This class is the very definition of a factory :), consider naming it as such.

bgavrilMS replied:
Please document all public classes and public methods, even for test code.

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs:
```diff
@@ -0,0 +1,34 @@
+            controller.EnterText("i0116", user.Upn, true);
+            controller.Tap("idSIButton9", true);
+            controller.EnterText("i0118", ((LabUser)user).GetPassword(), true);
+            controller.Tap("idSIButton9", true);
```
Do you control these IDs? If so, can you rename them to something more human readable? if not, can you add comments on what each point at? idSIButton9 is not very readable :)

trwalke replied:
The ones that are readable are controllable by me, the others are not. ill add comments.

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace Test.ADAL.NET.UIAutomation
+{
+    public static class CertificateHelper
```
It would help reviewers if you mentioned that you've copied code over from an existing repo, as we wouldn't have to review this again. You can add comments to your own PR. #Resolved

trwalke replied:
@bgavrilMS good point, ill clearly indicate that next time. thanks

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -0,0 +1,98 @@
+
+        private readonly string keyVaultClientID = "ebe49c8f-61de-4357-9194-7a786f6402b4";
+
+        private readonly string keyVaultThumbPrint = "440A5BE6C4BE2FF02A0ADBED1AAA43D6CF12E269";
```
You might want to run our internal compliance build before commiting, I think it might fail as it will consider there to be credentials.

bgavrilMS reviewed (approved):
Added some minor comments, bug LGTM otherwise. Please fix and commit.

bgavrilMS committed:
Merge branch 'dev' into trwalke/UiAutomationForMobile

Commit:
Modified MSAL test app to facilitate UI testing.
Added acquire token test for MSAL app.
Added comments.
Moving UI Test infrastructure to core.
Fixing Spelling.

trwalke committed:
Merge branch 'dev' into trwalke/UiAutomationForMobile

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/AppFactory.cs:
```diff
@@ -0,0 +1,25 @@
+                case Platform.iOS:
+                    return ConfigureApp.iOS.StartApp();
+                default:
+                    return ConfigureApp.Android.InstalledApp("com.Microsoft.XFormsDroid.ADAL").StartApp();
```
> return ConfigureApp.Android.InstalledApp [](start = 19, length = 41)

IMO: this should throw an exception saying platform not supported

trwalke replied:
@henrik-me I agree

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs:
```diff
@@ -0,0 +1,48 @@
+        [SetUp]
+        public void BeforeEachTest()
+        {
+            app = AppFactory.StartApp(platform);
```
>app [](start = 12, length = 3)

are you sure you want to re-create the app with-out setting it in the controller? This makes very little sense to me. Would be great if you can make the code reveal the intent

trwalke replied:
this is a bug, resolved

henrik-me commented on msal/tests/dev apps/XForms/XForms/AcquirePage.xaml:
```diff
@@ -65,14 +65,17 @@
-                <Button  Text="Clear"  Clicked="OnClearClicked" TextColor="Blue" Font="Bold,15"/>
+                <StackLayout Orientation="Horizontal"  HorizontalOptions="FillAndExpand">
+                    <Button  Text="Clear Result"  Clicked="OnClearClicked" TextColor="Blue" Font="Bold,15"/>
+                    <Button  Text="Clear Cache"  Clicked="OnClearCacheClicked" TextColor="Blue" Font="Bold,15" AutomationId="clearCacheButton"/>
```
any reason you are not setting AutomationId on all elements?

trwalke replied:
The UI framework does not seem to work well with the tabbed page view that is used for the MSAL app. it can only access the front page thus, I will add IDs for all elements on the front page. We will most likely need to build an application that is dedicated to UI testing and is the same across both libraries.

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/AppFactory.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// Initializes the app object that represents the main gateway to interact with the app on the device
+    /// </summary>
+	public class AppFactory
```
>AppFactory [](start = 14, length = 10)

any way this can be shared between ADAL and MSAL ... e.g. by a shared test core?

seems like we would like to reuse as much as possible including the actual tests. 

thoughts?

trwalke replied:
Yes, I can do that

Commit:
Moving AppFactory to core test project.

Added AutomationIds to all elements on the MSAL test app acquire page.

Added automationIds to all elements on the ADAL test app.

Added exception to appFactory for non supported platforms.

Fixed bug with app not being reset in the xamarinUITestController before each test.

henrik-me commented on core/tests/Test.Microsoft.Identity.Core.UIAutomation/AppFactory.cs:
```diff
@@ -0,0 +1,25 @@
+                case Platform.iOS:
+                    return ConfigureApp.iOS.StartApp();
+                default:
+                    throw new Exception("Platform not supported.");
```
>throw new Exception("Platform not supported."); [](start = 20, length = 47)

Pls. create a specific exception for this. e.g. PlatformNotSupportedException

henrik-me commented:
Looks really good, great refactorings. Only one minor comment remaining from my side.

Commit:
Added PlatformNotSupportedException class.

bgavrilMS reviewed (approved):
LGTM

trwalke committed:
Merge branch 'dev' into trwalke/UiAutomationForMobile

trwalke committed:
Merge branch 'dev' into trwalke/UiAutomationForMobile
